### PR TITLE
fix: rust coverage test meets compile error for missing debuginfo

### LINKED ISSUE #1652: rust coverage test could not compile error
When I tried to do coverage test with rust nightly-2023-02-09, I got

```
sccache: encountered fatal error
sccache: error: failed to zip up compiler outputs
sccache: caused by: failed to zip up compiler outputs
sccache: caused by: failed to open file `"/Users/yihau/yihau/sccache-coverage-error/target/debug/deps/unicode_ident-25938e1fedca0d89.gcno"`
sccache: caused by: No such file or directory (os error 2)
error: could not compile `unicode-ident`

Caused by:
  process didn't exit successfully: `/Users/yihau/.cargo/bin/sccache rustc --crate-name unicode_ident --edition=2018 /Users/yihau/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-ident-1.0.8/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=156 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C split-debuginfo=unpacked -C metadata=25938e1fedca0d89 -C extra-filename=-25938e1fedca0d89 --out-dir /Users/yihau/yihau/sccache-coverage-error/target/debug/deps -L dependency=/Users/yihau/yihau/sccache-coverage-error/target/debug/deps --cap-lints allow -Zprofile -Ccodegen-units=1 -Copt-level=0 -Clink-dead-code -Coverflow-checks=off -Zpanic_abort_tests -Cpanic=abort` (exit status: 254)
warning: build failed, waiting for other jobs to finish...
```

I have a tiny project can reproduce the error: https://github.com/yihau/sccache-coverage-build-error (The step is in the `coverage.sh`)

btw, It works fine with `nightly-2023-02-08`

### PR DESCRIPTION
**Pull request by yihau**
Hi @Xuanwo, does this one look correct?

Close #1652

### CODE REVIEW

**Comment by sylvestre** [comment]
I would like to see a test for that please. Thanks

**Comment by Xuanwo** [comment]
I don't know how `optional` works, maybe it ok for us to ignore all errors happened while writing cache?

**Comment by yihau** [comment]
I found the feature was introduced cuz `.dwo` files may absent.

https://github.com/mozilla/sccache/blob/main/src/compiler/gcc.rs#L536-L542

in the same PR, the author try to keep `.gcno` having the original behavior I think. so he make: 

https://github.com/mozilla/sccache/blob/main/src/compiler/gcc.rs#L550-L556

tbh I have no idea how sccache deal with those `.gcon` files. maybe .gcno are also optional files?

**Comment by yihau** [comment]
yeah. I think `.gcno` may absent as well in the newer nightly rust version.

I found that:

in `nightly-2023-02-08`, all deps files including `.d`, `.gcda`, `.gcno` and `-cgu.0.rcgu.o`

in `nightly-2023-02-09`, some of deps only have `.d` file

File: src/compiler/rust.rs [hunk 1]
```diff
@@ -186,7 +186,7 @@ pub struct RustCompilation {
     /// The compiler inputs.
     inputs: Vec<PathBuf>,
     /// The compiler outputs.
-    outputs: HashMap<String, PathBuf>,
+    outputs: HashMap<String, ArtifactDescriptor>,
```
Comment (yihau) [inline src/compiler/rust.rs:189]: make rust complier use optional outputs

File: src/compiler/rust.rs [hunk 2]
```diff
@@ -1513,19 +1513,37 @@ where
...
         if let Some(gcno) = gcno {
             let p = output_dir.join(&gcno);
-            outputs.insert(gcno.to_string_lossy().into_owned(), p);
+            outputs.insert(
+                gcno.to_string_lossy().into_owned(),
+                ArtifactDescriptor {
+                    path: p.clone(),
+                    optional: true,
+                },
+            );
```
Comment (yihau) [inline src/compiler/rust.rs:1546]: only make gcno optional. the others are not.

**Review by yihau — COMMENTED** [review]
copied the mechanism from c complier

**Review by Xuanwo — APPROVED** [review]
Mostly LGTM, thanks a lot!

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -609,3 +609,32 @@ jobs:
           ${SCCACHE_PATH} --show-stats
 
           ${SCCACHE_PATH} --show-stats | grep -e "Cache hits\s*[1-9]"
+
+  issues-1652:
```
Comment (sylvestre) [inline .github/workflows/integration-tests.yml:613]: Could you please rename the test for something more explicit?

File: .github/workflows/integration-tests.yml [hunk 4]
```diff
@@ -609,3 +609,32 @@ jobs:
...
+    runs-on: ubuntu-latest
+    needs: build
+
+    strategy:
+      fail-fast: false
+      matrix:
+        version: [nightly-2023-02-08, nightly-2023-02-09]
```
Comment (sylvestre) [inline .github/workflows/integration-tests.yml:620]: Is it necessary to hard-coded the date?

File: .github/workflows/integration-tests.yml (hunk 3 above)
Comment (yihau) [reply .github/workflows/integration-tests.yml:613]: do you have any prefer name? is missing-debug-info a good name?

File: .github/workflows/integration-tests.yml (hunk 4 above)
Comment (yihau) [reply .github/workflows/integration-tests.yml:620]: I found it's broken since the version `nightly-2023-02-09` so I choose those two version to test.

**Comment by sylvestre** [comment]
I would prefer that you extend it to test coverage in general

**Comment by yihau** [comment]
oh. do you mean that 1) change job name to `test coverage` and 2) use the latest nightly version ?

**Comment by yihau** [comment]
Hi @sylvestre, I have a diff in my local. it looks like:
```diff
diff --git a/.github/workflows/integration-tests.yml b/.github/workflows/integration-tests.yml
index cf3e909..3ad4cbe 100644
--- a/.github/workflows/integration-tests.yml
+++ b/.github/workflows/integration-tests.yml
@@ -610,14 +610,9 @@ jobs:

           ${SCCACHE_PATH} --show-stats | grep -e "Cache hits\s*[1-9]"

-  issues-1652:
+  rust-test-coverage:
     runs-on: ubuntu-latest
     needs: build
-
-    strategy:
-      fail-fast: false
-      matrix:
-        version: [nightly-2023-02-08, nightly-2023-02-09]
     steps:
       - uses: actions/download-artifact@v3
         with:
@@ -628,11 +623,11 @@ jobs:

       - name: Coverage test
         run: |
-          rustup toolchain install ${{ matrix.version }}
+          rustup toolchain install nightly
           cargo new hello
           cd hello
           echo "serde = { version = \"1.0\", features = [\"derive\"] }" >> Cargo.toml
-          cargo +${{ matrix.version }} test
+          cargo +nightly test
         env:
           RUSTC_WRAPPER: /home/runner/.cargo/bin/sccache
           CARGO_INCREMENTAL: "0"
```

I will push this code if you think it is better

**Comment by sylvestre** [comment]
yeah, exactly what I was expecting :)

File: .github/workflows/integration-tests.yml [hunk 5]
```diff
@@ -609,3 +609,27 @@ jobs:
...
+      - name: Chmod for binary
+        run: chmod +x ${SCCACHE_PATH}
+
+      - name: Coverage test
+        run: |
+          rustup toolchain install nightly
+          cargo new hello
+          cd hello
+          echo "serde = { version = \"1.0\", features = [\"derive\"] }" >> Cargo.toml
+          cargo +nightly test
```
Comment (sylvestre) [inline .github/workflows/integration-tests.yml:630]: can you please add a check that we are indeed caching the right things?

see the other integration tests

Comment (yihau) [reply .github/workflows/integration-tests.yml]: sorry for the incorrect path. fixed it in [10b37a7](https://github.com/mozilla/sccache/pull/1740/commits/10b37a7c4ea304d40577c21c35814b936146f5e5)

and check it in my fork repo. it succeed
Screenshot 2023-04-21 at 6 30 08 PM

**Comment by Xuanwo** [comment]
This PR seems ready to go now. @sylvestre Do you wanna to take another look?

**Comment by ryoqun** [comment]
hey, it looks like [v0.4.2](https://github.com/mozilla/sccache/releases/tag/v0.4.2) has shipped yesterday and includes this pr. really thanks for it. but i'm wondering why v0.4.2 isn't still appearing at [crates.io](https://crates.io/crates/sccache/versions)? (we use that release channel for convenience as part of our docker ci image generation).

**Comment by Xuanwo** [comment]
seems we didn't push to crates.io in our release workflow.

**Pull request merged**